docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,403 @@
+# Best Practices for Keeping Enclaves Secure
+Use this guide to apply secure patterns and avoid common mistakes that put the
+security of enclave applications at risk.
+
+1.  [Best Practices for Interface Custom Marshaling ](#1-best-practices-for-interface-custom-marshaling)
+
+    1.1 [Ensuring memory is where it should be](#11-ensuring-memory-is-where-it-should-be)
+
+    1.2 [SGX enclaves: oe_is_within_enclave vs. oe_is_outside_enclave](#12-sgx-enclaves-oe_is_within_enclave-vs-oe_is_outside_enclave)
+
+    1.3 [TOCTOU or double fetch vulnerabilities](#13-toctou-or-double-fetch-vulnerabilities)
+
+    1.4 [Ensure proper bounding of data buffers](#14-ensure-proper-bounding-of-data-buffers)
+
+    1.5 [Corrected enclave ecall with custom marshaling](#15-corrected-enclave-ecall-with-custom-marshaling)
+
+2.  [Handling Secrets in Enclave Applications](#2-handling-secrets-in-enclave-applications)
+
+    2.1 [How _Not_ to Handle Application Secrets](#21-how-not-to-handle-application-secrets)
+
+    2.2 [The New Secure Enclave Way to Handle Application Secrets](#22-the-new-secure-enclave-way-to-handle-application-secrets)
+
+<br />
+<br />
+<br />
+
+# 1. Best Practices for Interface Custom Marshaling 
+(For an overview of using the Enclave Definition Language (EDL) and the
+oeedger8r tool to produce enclave interface code, please refer to
+[Getting started with the Open Enclave edger8r](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/Edger8rGettingStarted.md).)
+
+Calling into and out of enclaves uses special methods that orchestrate the
+context switch and marshal the function parameters. Much of the code necessary
+to manage these calls and parameter marshaling are common to most applications.
+The Open Enclave _oeedger8r_, using the interface definition defined by
+application developers in EDL files, generates boilerplate code for them.
+
+In some uncommon cases, developers may want to pass data types that are not
+defined at the interface level or handle marshaling differently from what the
+oeedger8r tool generates. To specify this, use `user_check` for the parameter
```

Comment:
```suggestion
oeedger8r tool generates. Although this is not generally recommended, it is supported.
To specify this, use `user_check` for the parameter
```

Suggested text is my proposal to make it clear that doing so is not being stated to be a "best practice", but if you DO do it, then this doc covers the best practice on HOW.  Line 61-62 already does so, but this suggestion makes it clear up front for consistency in message.